docs/Training-Generalization-Learning.md

```diff
@@ -0,0 +1,105 @@
+
+Ball scale of 0.5          |  Ball scale of 4
+:-------------------------:|:-------------------------:
+![](images/3dball_small.png)  |  ![](images/3dball_big.png)
+
+_Variations of the 3D Ball environment._
+
+To vary the environments, we first decided what the reset parameters for the environment. These parameters are known as `Reset Parameters`, introduced in [Curriculum Learning](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-Curriculum-Learning.md). In the 3D ball environment example displayed in the figure above, the reset parameters are `gravity`, `ball_mass` and `ball_scale`.
+
+## In-practice
+
+To test the effectiveness of this training procedure, we train 3 models over 50000 steps:
+    1. Model Trained on Default reset parameter values (Default model).
+    2. Model Trained on a range of reset parameter values (Random model). Reset parameter values are picked uniformly over the range and the model is trained on each configuration for 5000 steps before they are randomly sampled again. The range consists of the default values.
+    3. Model Trained on a range of reset parameters values (Extreme model). The range here are the extreme ends, which lie outside the range used to train the Random model. Reset parameter values are selected randomly and the model is trained for 5000 steps on each configuration of values before resampling the values again.
+
+Then, these 3 models are tested for their performance in all the reset parameter settings for 50000 steps each. The figure below outlines the performance of each model.
+
+![3DBall environment mean rewards](images/3DBall_generalization_rewards.png)
+
+In a relatively simple environment we are able to see differences in the performance between differently trained models, with the models exposed to various environments performing better in variations of environments than the default model trained over a single environment.
+
+## How-to
+
+For generalization training, we need to provide a way to modify the environment by supplying a set of reset parameters. This provision can be either deterministic or randomized. Each reset parameter is assigned a sampler. If a sampler isn't provided for a reset parameter, the parameter maintains the default value throughout the training, remaining unchanged. The samplers for all the reset parameters are handled by a **Sampler Manager**, which is also responsible for generating a new set of values for the reset parameters when needed. 
+
+To setup the Sampler Manager, we setup a YAML file that specifies how we wish to generate new samples. In this file, we specify the samplers and the `resampling-duration` (number of training steps after which reset parameters are resampled). Below is an example of a sampler file for the 3D ball environment.
```

Comment:
`number of training steps after which reset parameters are resampled` --> these are not necessarily training steps, they are just simulations steps.